rust/sdk/src/client.rs

```diff
@@ -388,6 +316,356 @@ pub(crate) fn to_lines(stream: impl Stream<Item = Result<Bytes, reqwest::Error>>
+                Publish::Empty => panic!("Polling a terminated Publish future"),
+            };
+        }
+    }
+}
+
+impl<'a> FusedFuture for Publish<'a> {
+    fn is_terminated(&self) -> bool {
+        if let Publish::Empty = self {
+            true
+        } else {
+            false
+        }
+    }
+}
+
+pub enum Query<'a> {
+    Initial {
+        client: &'a ActyxClient,
+        request: QueryRequest,
+    },
+    Pending(BoxFuture<'a, anyhow::Result<BoxStream<'a, QueryResponse>>>),
+    Empty,
+}
+
+impl<'a> Query<'a> {
+    fn new<Q: Into<String>>(client: &'a ActyxClient, query: Q) -> Self {
+        Self::Initial {
+            client,
+            request: QueryRequest {
+                query: query.into(),
+                lower_bound: Some(OffsetMap::empty()),
+                upper_bound: None,
+                order: Order::Asc,
+            },
+        }
+    }
+
+    /// Add a lower bound to the query.
+    ///
+    /// The lower bound limits the start of the query events.
+    /// As an example, consider the following (example) events:
+    /// ```json
+    /// { "lamport": 1, "event": { "temperature": 10 } }
+    /// { "lamport": 3, "event": { "temperature": 12 } }
+    /// { "lamport": 14, "event": { "temperature": 9 } }
+    /// ```
+    /// If you set the lower bound to `10`, only the last event will be returned.
+    pub fn with_lower_bound(mut self, lower_bound: OffsetMap) -> Self {
+        if let Self::Initial { ref mut request, .. } = self {
+            request.lower_bound = Some(lower_bound);
+        }
+        self
+    }
+
+    /// Add an upper bound to the query.
+    ///
+    /// The upper bound limits the start of the query events.
+    /// As an example, consider the following (example) events:
+    /// ```json
+    /// { "lamport": 1, "event": { "temperature": 10 } }
+    /// { "lamport": 3, "event": { "temperature": 12 } }
+    /// { "lamport": 14, "event": { "temperature": 9 } }
+    /// ```
+    /// If you set the upper bound to `10`, the first two events will be returned.
+    pub fn with_upper_bound(mut self, upper_bound: OffsetMap) -> Self {
+        if let Self::Initial { ref mut request, .. } = self {
+            request.upper_bound = Some(upper_bound);
+        }
+        self
+    }
+
+    /// Set the query's event order. By default, this value is set to [`Order::Asc`].
+    ///
+    /// As an example, consider the following (example) events:
+    /// ```json
+    /// { "lamport": 1, "event": { "temperature": 10 } }
+    /// { "lamport": 3, "event": { "temperature": 12 } }
+    /// { "lamport": 14, "event": { "temperature": 9 } }
+    /// ```
+    /// If your query sets [`Order::Desc`], the result will instead look like:
+    /// ```json
+    /// { "lamport": 14, "event": { "temperature": 9 } }
+    /// { "lamport": 3, "event": { "temperature": 12 } }
+    /// { "lamport": 1, "event": { "temperature": 10 } }
+    /// ```
+    pub fn with_order(mut self, order: Order) -> Self {
```

Comment:
What happens if I pass this query `FROM 'sometag' ORDER DESC` and then set `with_order(Order::ASC)`?

Reply:
https://github.com/Actyx/Actyx/blob/7685fe6d1fba2f577cc804b3b60831d71d843418/rust/actyx/api/src/events/service.rs#L139

The precedence is:
1. from AQL
2. inferred through AGGREGATE
3. from the API request

Reply:
Great question, I've taken this into account for the next revision